tests/lib/fortigate_test.py

```diff
@@ -0,0 +1,496 @@
+# Copyright 2019 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unittest for fortigate policy rendering module."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import re
+import string
+import unittest
+import mock
+
+from capirca.lib import fortigate
+from capirca.lib import nacaddr
+from capirca.lib import naming
+from capirca.lib import policy
+
+
+GOOD_HEADER = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate from-id 2
+}
+"""
+
+GOOD_HEADER_1 = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate ngfw-mode policy-based
+}
+"""
+
+BAD_HEADER = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate edge-filter
+}
+"""
+
+TERM_TEMPLATE = """
+term good-term-2 {{
+  source-interface:: {src_interface}
+  destination-interface:: {dest_interface}
+  protocol:: {protocol}
+  destination-address:: {dest_addr}
+  destination-port:: {dest_port}
+  source-address:: {src_addr}
+  source-port:: {src_port}
+  action:: {action}
+  logging:: {logging}
+}}
+"""
+
+_SP = '    '
+
+EXP_INFO = 2
+
+
+class CustomFormatter(string.Formatter):
+  """
+  Checks the custom formatter for fortigate output.
+
+  """
+  DEFAULT_VALUES = {
+      'src_interface': 'wan1',
+      'dest_interface': 'wan2',
+      'protocol': 'tcp',
+      'src_addr': 'SOME_HOST',
+      'dest_addr': 'SOME_HOST',
+      'src_port': 'HTTP',
+      'dest_port': 'HTTP',
+      'action': 'accept',
+      'logging': 'true'
+  }
+
+  def format(*args, **kwargs):
+    if 'remove_fields' in kwargs or 'add_fields' in kwargs:
+      args = list(args)
+
+    if 'remove_fields' in kwargs:
+      for field in kwargs['remove_fields']:
+        remove_regex = '.*' + field + '.*'
+        args[1] = re.sub(remove_regex, '', args[1])
+
+    if 'add_fields' in kwargs:
+      add_fields_string = ""
+      for field, value in kwargs['add_fields'].items():
+        add_fields_string += "  " + field + ":: " + value + "\n"
+      args[1] = args[1][:-3] + add_fields_string + args[1][-3:]
+
+    return string.Formatter.format(*args, **kwargs)
+
+  def get_value(self, key, args, kwds):
+    try:
+      return kwds[key]
+    except KeyError:
+      return self.DEFAULT_VALUES[key]
+
+
+class FortigateTest(unittest.TestCase):
```

Comment:
Could we please add a testcase (a follow-up is fine) for

Validating ICMP and ICMPv6 are correctly rendered for IPv4 and IPv6 terms.